### ISSUE TITLE
Text kerning is not working

### ISSUE THREAD

**Issue by oomek** [SFML/SFML#1642]
## Subject of the issue

It seems like all modern fonts that use GPOS table for storing kerning tables are not displayed properly in SFML. The only fonts that have proper kerning applied are those with KERN table in "legacy" format.

## Your environment

* Windows 10
* git master
* WSL Ubuntu x-compiled with MXE packages
* Special compiler flags used: none

## Steps to reproduce

Download a modern ttf font like for example Barlow:
https://github.com/jpt/barlow
Draw in SFML and in Photoshop
Compare kerning of pairs with high kern values like:
Vo, Wo, To, VA, AV

## Expected behavior

Kerning should be applied like in Photoshop

## Actual behavior

Kerning not applied when the kerning table is stored in GPOS table

![image](https://user-images.githubusercontent.com/2974860/74931356-578aa700-53d7-11ea-935f-5e2e1f9bf0b7.png)

**Comment by LaurentGomila** [comment]
For font rendering issues you should always first check the FreeType documentation, as we highly rely on it.

> FreeType only supports kerning via the (rather simple) ‘kern’ table. For the interpretation of kerning data in the (highly sophisticated) ‘GPOS’ table you need a higher-level library like ICU or HarfBuzz since it can be context dependent (this is, the kerning may vary depending on the position within a text string, for example).

**Comment by oomek** [comment]
Noted.

**Comment by oomek** [comment]
Would you accept a pr adding harfbuzz for glyph advance calculations?

**Comment by LaurentGomila** [comment]
If using Harfbuzz, it would then be for the whole implementation of sf::Text, no need to limit to just this specific bit.

But it's quite a big dependency, I don't know if we really want that.

**Comment by oomek** [comment]
I'm not sure it's even possible. Harfbuzz does not render glyphs as far as I know. I see it rather as they're complementary. HB even takes FT font instance as a parameter. I'm willing to try to add a proper kerning to SFML, we just need to find some compromise features/complexity beforehand.

**Comment by LaurentGomila** [comment]
I never talked about rendering glyphs, sf::Text doesn't do that. As I see it, it would be sf::Text == Harfbuzz and sf::Font == FreeType.

**Comment by oomek** [comment]
Yeah, but getKerning() function lives in font.cpp doesn't it ;) But I get what you mean, HarfBuzz calls would reside in sf::Text

**Comment by eXpl0it3r** [comment]
We already had a HarfBuzz PR/draft once. I think in the long run it's a must, as the current implementation can't handle complex text layouts either. But rather than adding new binaries to the repo, we should ideally make a final decision on [how to handle dependencies](https://en.sfml-dev.org/forums/index.php?topic=25053.0) - still need to finish my solution design.

**Comment by oomek** [comment]
Is this draft still available on github somewhere?

**Comment by eXpl0it3r** [comment]
Yes #624

**Comment by oomek** [comment]
There is one more thing that should be addressed. SFML uses forced hinting so the small characters look crisper, but this according to a not in freetype's website requires compensation. Otherwise small glyphs are too wide apart. I've added what they suggested and it looks way better. Since it's quite drastic change on small glyphs do you think it should be default or optional? Freetype says positioning auto-hinted glyphs without compensation is the wrong way.

![compensation](https://user-images.githubusercontent.com/2974860/75612597-fa040200-5b1c-11ea-951d-e203622a5d86.gif)

**Comment by oomek** [comment]
Here is another example using segoeui. It's not as severe when the font is not narrow and bold as in the Barlow Condensed, but it fixes the uniformity of the glyph spacing.

![compensation-segoeui](https://user-images.githubusercontent.com/2974860/75612920-fde55380-5b1f-11ea-9ffb-224b0f2e3689.gif)

**Comment by oomek** [comment]
The link to the note below ( second pseudocode block ):
https://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_glyphslotrec

I've applied it in the `ensureGeometryUpdate()` in `Text.cpp` after `// Extract the current glyph's description`

**Comment by eXpl0it3r** [comment]
I think it makes sense to use this by default. Might cause some alignment issues for those who relied on some fixed positioning or widths, other than that, I think it does look better and if FT says it's the way to go, then we should also apply this.

But maybe @LaurentGomila has some insights on that topic.

**Comment by LaurentGomila** [comment]
Nothing to add, it looks better and FreeType advises it, so let's go :+1:

**Comment by oomek** [comment]
Awesome. There is ine last thing I would like to discuss. I use the formula below to calculate the offset:
`x += std::floor((glyph.lsb_delta - prevRsbDelta + kerning + 31) / 64.0);`
This way I quantize the offset with kerning and delta combined which gives the best uniformity.
The only problem is that `GetKerning()` function returns an int and already divided by 64. I also changed the flag to `FT_KERNING_UNFITTED` to skip quantization stage in `FT_Get_Kerning()` call.

Since `GetKerning()` is a public function what should I do? Make another function that returns a raw kerning, or add andadditional parameter that changes the behaviour of mentioned function?

**Comment by LaurentGomila** [comment]
The code in FreeType documentation looks simpler than yours: kerning remains separated, and it doesn't use rsb_delta from previous glyph to render the current one.

Why did you implement it differently?

**Comment by oomek** [comment]
It does. I used the snippet below and got rid of branching.
```
    if ( prev_rsb_delta - slot->lsb_delta >  32 )
      origin_x -= 64;
    else if ( prev_rsb_delta - slot->lsb_delta < -31 )
      origin_x += 64;

    prev_rsb_delta = slot->rsb_delta;
```

. My oneliner serves the same function. I added kerning at this stage to avoid adding 2 separately rounded offsets.

**Comment by oomek** [comment]
The compensation offset is never larger than 64 so it was safe to strip the branching off.
Imagine you have a compensation of hinting 30 and kerning 30 in font units. When you round them separately the final offset will be 0 pixels

**Comment by LaurentGomila** [comment]
I was referring to the other piece of code; they say this one is for integer positionning. Which is *not* what we do, all coordinates are converted from FT fixed point to standard floating point. So I don't get what you mean when talking about rounding, integer coordinates, etc.

**Comment by oomek** [comment]
I need to make sure we are on the same page. I'm talking about generating a vertex buffer for strings which is fixed to integers, not the the float transforms and positioning. Atlas based fonts will always look blurry when you do not stick to integers when transforming pos and size. The only solution for it is SDF, but that's another topic on which I'm working on. It would not be a replacement for the existing font class, but it would allow to scale and position sharp fonts with float precision.

**Comment by oomek** [comment]
There is also a glyph spacing variable which if not equal to 1/3 of space size will also result in blurred fonts. This fix does not address this issues, only proper glyph spacing.

**Comment by LaurentGomila** [comment]
> I need to make sure we are on the same page.

Agreed 😄

> I'm talking about generating a vertex buffer for strings

Ok, `sf::Text::ensureGeometryUpdate()`, as you said previously.

> which is fixed to integers

I don't remember all the details of the implementation, but after a quick check I couldn't find the place where this rounding to integer happens. Can you show me?

> The only solution for it is SDF, but that's another topic on which I'm working on

:+1:

**Comment by oomek** [comment]
What I meant is there are integers stored in floats. Width and height of quads is based on generated bitmaps, so it's integers. Kerning and advance is also quantized to pixels. FT_KERNING_DEFAULT gives multiplies of 64 when I logged the output of GetKerning() function.

**Comment by oomek** [comment]
Here's also a note regarding horiAdvance:
`
linearHoriAdvance - This field contains the linearly scaled value of the glyph's horizontal advance width. Indeed, the value of metrics.horiAdvance that is returned in the glyph slot is normally rounded to integer pixel coordinates (i.e., it will be a multiple of 64) by the font driver used to load the glyph image. linearHoriAdvance is a 16.16 fixed float number that gives the value of the original glyph advance width in 1/65536th of pixels. It can be use to perform pseudo device-independent text layouts.
`

**Comment by LaurentGomila** [comment]
> Kerning and advance is also quantized to pixels.

Ah, that's it. Ok for advance, and is it documented for kerning as well?

As I said I don't remember every detail, but in my opinion the "right" way to handle coordinates would be to keep x/y positions as floating points, and only round coordinates to integer when adding quads to the vertex array. So sf::Font functions should not return rounded values if possible. What do you think?

**Comment by oomek** [comment]
Yes, it's also doccumented

The default value is FT_KERNING_DEFAULT, which has value 0. It corresponds to kerning distances expressed in 26.6 grid-fitted pixels (which means that the values are multiples of 64). For scalable formats, this means that the design kerning distance is scaled, then rounded.

That's why I changed the flag to get not rounded value which is then combined with not rounded bearing deltas and then it's rounded explicitly.

I could try to move that line to GetKerning() function. I'll compare the screenshots and get back to you in the evening.

**Comment by oomek** [comment]
Hi. I've had an internet outage in my area for several days, but I'm back. I've been studying the font class in the meantime and I've stumbled on a tiny roadblock.  Let me explain.

The goal is to store the `lsb_delta` and `rsb_delta` in the Glyph class table. I would like to access it in `getKerning()`  like it's accessed in `getGlyph()` function , but there is a catch. I do not know whether the font is bold or has an outline, which is not important for applying deltas as the extra space is already contained in `advance` field, but I need to somehow retrieve any glyph data for a certain size regardless of any extra flags ( bold, outline )

Is there any way of applying a mask to the `find` function so it returns a glyph without generating a new one. This would happen if I was drawing bold glyphs and just called `getGlyph()` with `bold=false`

```
    GlyphTable::const_iterator it = glyphs.find(key);
    if (it != glyphs.end())
    {
        // Found: just return it
        return it->second;
    }
```

**Comment by LaurentGomila** [comment]
> I would like access that table it in getKerning()
> [...]
> I need to somehow retrieve any glyph data for a certain size regardless of any extra flags ( bold, outline )

Why? Can you explain what you're trying to do?

**Comment by oomek** [comment]
You said you would like me to move deltas calculation from `Text` to `Glyph` class, so I thought the best way would be to make `getKerning()` function returning the offset with already hinting deltas applied, but I do not have an access to style flags there and I did not want to add any additional parameters to it as it's an exposed function.

**Comment by oomek** [comment]
Nevermind, its all in `m_info` and can be accessed with `getInfo()` function 😅

**Comment by LaurentGomila** [comment]
Ah, I didn't get that you were trying to merge all that stuff with kerning. It's a different thing so I'm fine if it's separate. For example, turn Glyph's advance into left/right margin (right would be current advance + the new value). Or something more clever -- I haven't thought about it more than 30 seconds.

m_info? What's there except the font family?

**Comment by oomek** [comment]
Bold and outline, which is needed to get the correct glyph from the table (combine function) which would store the hinting deltas. Putting it into kerning is the best solution in my opinion as I need to do the calculations before dividing by 64 it will also work when getkerning is used alongside with findCharacterPos to get a precise position inbetween letters if you want to draw a blinking curasr for example.

**Comment by LaurentGomila** [comment]
> Bold and outline

Are we looking at the same source code? 😃
Bold and outline are per glyph while the Info structure is per font.

> I need to do the calculations before dividing by 64

I don't think there's any loss when switching from fixed point to floating point. So if needed it could perfectly be done *after* dividing by 64.

If you can put that stuff in kerning that's fine, that would allow to get the improved layout without any modification to sf::Font API and to sf::Text implementation.

**Comment by oomek** [comment]
> Are we looking at the same source code? 😃
Apparently not. I keep catching myself reading the comment blocks BELOW the function and I read the description of getGlyph() :D damn it, back to the beginning. 
> I don't think there's any loss when switching from fixed point to floating point. So if needed it could perfectly be done after dividing by 64.
Since I use a `FT_KERNING_UNFITTED` flag I need to divide the kerning by 64 and then floor it before returning the value in pixels, and by doing this I'm left with the fractional part stripped off. `FT_KERNING_DEFAULT` quantizes the kerning in 64 unit increments.

Maybe I just push the commit with a comment where I'm struggling so we have a base to further talks.

**Comment by LaurentGomila** [comment]
> and then floor it

Why? It should be fine to return a non-integer value. As I said, we should rather round coordinates at the final (rendering) step, and work with accurate coordinates internally.

> Maybe I just push the commit with a comment where I'm struggling so we have a base to further talks.

You can do this, yes.

**Comment by oomek** [comment]
> Why?

`FT_KERNING_UNFITTED` flag requires it.

https://github.com/SFML/SFML/compare/master...oomek:bugfix/autohint_compensation

**Comment by oomek** [comment]
Pushed a second commit with a fix that I've missed.
https://github.com/oomek/SFML/commits/bugfix/autohint_compensation

**Comment by LaurentGomila** [comment]
> FT_KERNING_UNFITTED flag requires it. We want the kerning returned in full pixels like it was before the change.

But for what reason do we want that?

**Comment by oomek** [comment]
Otherwise you will get blurred fonts caused by subpixel positioning of autohinted glyphs

**Comment by oomek** [comment]
![image](https://user-images.githubusercontent.com/2974860/76168125-6ce82b00-6164-11ea-9b73-9d1d53521d7f.png)

**Comment by LaurentGomila** [comment]
As I said, I really think that rounding should happen in a final stage, when definining quad coordinates, rather than in each glyph layout attribute.

**Comment by oomek** [comment]
> As I said, I really think that rounding should happen in a final stage, when definining quad coordinates, rather than in each glyph layout attribute

Could you tell me why please?

If we gonna make rounding after:
`x += glyph.advance + letterSpacing;`
it's gonna break letter spacing and cause "independent glyphs crawling" ( they will stick to integer positions randomly) when you try to animate it. No one sane would use letter spacing other than 1.0 for very small glyphs anyway. On bigger ones the blur is not visible and it looks nice when letters smoothly go apart, so we should keep letter spacing as a float.

**Comment by jkeller51** [comment]
@oomek I just wanted to say. I was struggling all day trying to fix the hinting spacing issue. I am working on a production app and I don't know anything about font rendering, but we have a very particular UI/UX guy who was complaining (with good reason) about the SFML-generated text spacing vs. the text in our bitmap assets. The best thing I could come up with was rendering the text at 4x the size and then scaling it 0.25x before drawing, but the result wasn't as clear as the native-resolution text. And it would have taken probably another whole day to actually implement this super-resolution change across the whole app.

Then I stumble across this issue, work out from your link how to modify SFML to incorporate the compensation... and the result is perfect! Beautiful!

You saved me a bunch of time and got me better results to boot. Thank you!

**Comment by oomek** [comment]
I'm glad it helped you. The issue is stuck though for some reason as we haven't come to a final conclusion yet.

**Comment by oomek** [comment]
This issue has been stuck in undecided state for 8 months. I would appreciate some further feedback on that, because I don't know whether I should keep working on it, or forget and close the issue.

**Comment by eXpl0it3r** [comment]
Sorry for the delay! 🙁 

I suggest to make a PR and then we can hammer out some finer details. For me as someone who's not very familiar with FreeType or SFML's font implementation, the conversation here, is rather hard to follow.

**Comment by mahuti** [comment]
I would love to see this implemented. It can make a huge difference especially on smaller screens. Crisper spacing and rendering makes these tiny fonts more legible, and sometimes they're the only option.

**Comment by jkeller51** [comment]
Okay, I will look back at my changes sometime this week and see about opening a PR. I agree that the changes were very important for me and my team.

**Comment by oomek** [comment]
My proposed changes are here.
https://github.com/oomek/SFML/commits/bugfix/autohint_compensation
Maybe as an adition autohinting should be optional? When on it will use my code, when off it's not needed

**Comment by eXpl0it3r** [comment]
Could you create a PR with those changes? 
It's a lot easier to have discussions directly in code, than to switch back and forth between code and discussion thread.

**Comment by jkeller51** [comment]
I came up with my changes independently last May, and it looks like some are different than yours @oomek . I have no doubt that you know what you're doing, but wanted to add my changes for comparison.

I think the biggest difference is @oomek's code compensates for the kerning in `sf::Font::getKerning()`, while I left that function as-is and did the compensation in `sf::Text::findCharacterPos()` and `sf::Text::ensureGeometryUpdate()`. @oomek 's factoring is maybe cleaner, but depending on how slow `sf::Font::getGlyph()` is, my version has the advantage of not retrieving the same glyph twice in a row. It cuts down calls to `getGlyph()` by half.

**Comment by oomek** [comment]
I'm gonna rebase and pr tomorrow if that's ok.

**Comment by oomek** [comment]
@jkeller51 I read the freetype docs and my commit is based on their guide. You have to use lsb and rsb and compensate the positioning on that if you set forced autohint flag.

**Comment by jkeller51** [comment]
@oomek yes, my changes were also based on the freetype docs, and I am using the lsb and rsb to compensate. Just in a different place.

**Comment by oomek** [comment]
Ok, remind me a link please I'm gonna try it on my testing app I used for screenshots

**Comment by jkeller51** [comment]
Try this: https://github.com/SFML/SFML/pull/1745

**Comment by oomek** [comment]
I've compared our both PRs by making a lot of screenshots of various fonts. There are only slight 1 pixel variations between certain kerning pairs and the space width. Some are bigger some smaller, nothing serious. Unfortunately your PR is suffering from GetKerning() function returning incorrect values. This was the main reason of me writing it that way. I've extensively discussed it in my previous posts.

**Comment by oomek** [comment]
A little update, after further testing. Both our commits are breaking my word wraping function but in a different manner. I'm gonna try to get to the bottom of this tomorrow.

**Comment by jkeller51** [comment]
That's correct, I left `getKerning()` as it was before. If `getGlyph()` is nontrivial, then it would be a simple matter to use your `getKerning()` for public access, and have an internal `getRawKerning()` with the old behavior, that is used by `sf::Text::findCharacterPos()` and `sf::Text::ensureGeometryUpdate()`, as in my commit. This would reduce calls to `getGlyph()` by half in those functions.

If `getGlyph()` is trivial, then it might be better to just use your changes entirely.

**Comment by oomek** [comment]
I've identified the problem with my PR. I left `FT_Vector kerning` uninitialized so the fix was the following:
```
FT_Vector kerning;
kerning.x = kerning.y = 0;
if( FT_HAS_KERNING(face))
    FT_Get_Kerning(face, index1, index2, FT_KERNING_UNFITTED, &kerning);
 ```
Now my wordwrap function is back again pixel perfect.
The reason of moving lsb/rsb to `getKerning()` function was to get rid of double integer rounding. I combine `kerning, lsb, rsb` and only after I bitshift and round. I strongly believe this is how it should be done.
I was also thinking about adding another internal function, but I also need to use proper `getKerning()` inside my app as this is how I calculate the length of the text so it's pixel perfect to the bounding box:
 ```
kerning = font->getKerning( s[std::max( 0, i - 1 )], s[i], charsize );
running_total += kerning;
g = &font->getGlyph( s[i], charsize, false );
running_width = std::max( running_width, (int)( running_total + g->bounds.left + g->bounds.width ));
running_total += g->advance;
```
I do not use findCaracterPos() because as you said way too many unnecessary calls.

**Comment by oomek** [comment]
I've updated my PR. Please tell me if this is acceptable:
https://github.com/SFML/SFML/blob/d2406fcff21947e4ad125f40f1234787b40b2cee/include/SFML/Graphics/Font.hpp#L220

**Comment by eXpl0it3r** [comment]
> if this is acceptable

With regards to the API, yes that's okay. It breaks ABI but not API and we care about ABI to some degree only for patch versions. 
As for the whole PR, I want Laurent to take a look, as I'm as familiar with Freetype or text rendering...

**Comment by eXpl0it3r** [comment]
@oomek now that the PR is merged, what exactly remains to be discussed next regarding the kerning?

**Comment by eXpl0it3r** [comment]
I'm closing this for now, we can always open a new issue to discuss further kerning improvements.
